rcl/src/rcl/node_type_cache.c

```diff
@@ -0,0 +1,240 @@
+// Copyright 2023 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rcl/node_type_cache.h"
+#include "rcl/type_description_conversions.h"
+
+#include "rcl/error_handling.h"
+#include "rcutils/logging_macros.h"
+#include "rcutils/types/hash_map.h"
+
+#include "./context_impl.h"
+#include "./node_impl.h"
+
+typedef struct rcl_type_info_with_registration_count_t
+{
+  /// Counter to keep track of registrations
+  size_t numRegistrations;
+
+  /// The actual type info.
+  rcl_type_info_t type_info;
+} rcl_type_info_with_registration_count_t;
+
+size_t get_type_hash_hashmap_key(const void * key)
+{
+  // Reinterpret-cast the first sizeof(size_t) bytes of the hash value
+  const rosidl_type_hash_t * type_hash = key;
+  return *(size_t *)type_hash->value;
```

Comment:
@emersonknapp can you comment on Chris' question? That's your magic code ;)

Reply:
Not magic, I promise!

The signature for the `rcutils_hash_map` key hashing function is `size_t (const void * key)`  - this is sensible because the hashmap is just an array of buckets, and an array is indexed/addressed by `size_t`, which yes is system-dependent in size. So that explains why it's a `size_t`, that's the required (eventual) hash type in order to use the `rcutils_hash_map`.

That being the constraint - our keys for the hash map are a RIHS01 value, which is just a SHA256 value. This is already a hash of real content, it would be a waste of computation to perform further hashing on it, it's as "random" as it could be for the information represented - this includes within the hash itself. None of the bits of a SHA256 hash should be considered privileged to any others in terms of information content, so for a `size_t` amount of uniqueness, we can just lop off a `size_t`-worth of the full 256 bits to use as our hash. The beginning, middle, or end should make no difference.

I'll note that I'm not so positive on the above paragraph that I'd use it without further study for security purposes, but in this case it's just for a simple lookup table so I expect on average no more collisions between "first N bytes of SHA256 value" vs "SHA256 put through some hashing function to output an N byte hash", and the former option requires far less resources.

Reply:
I see, that makes sense.  Thanks for the explanation!